### ISSUE TITLE
Cannot include Quantities in FITS headers

### ISSUE THREAD

**Issue by parejkoj** [astropy/astropy#9332]
It looks like one cannot put an astropy Quantity into a FITS header (astropy version=`3.1.2`):

```
import astropy.units as u
from astropy.io import fits
temperature = 10*u.deg_C
header = fits.Header()
header['CCDTEMP'] = temperature
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-7-3ce5059bc40e> in <module>
----> 1 header['CCDTEMP'] = temperature

~/lsst/lsstsw/miniconda/envs/lsst-scipipe/lib/python3.7/site-packages/astropy/io/fits/header.py in __setitem__(self, key, value)
    187             # If we get an IndexError that should be raised; we don't allow
    188             # assignment to non-existing indices
--> 189             self._update((key, value, comment))
    190 
    191     def __delitem__(self, key):

~/lsst/lsstsw/miniconda/envs/lsst-scipipe/lib/python3.7/site-packages/astropy/io/fits/header.py in _update(self, card)
   1607         else:
   1608             # A new keyword! self.append() will handle updating _modified
-> 1609             self.append(card)
   1610 
   1611     def _cardindex(self, key):

~/lsst/lsstsw/miniconda/envs/lsst-scipipe/lib/python3.7/site-packages/astropy/io/fits/header.py in append(self, card, useblanks, bottom, end)
   1139             card = Card(card)
   1140         elif isinstance(card, tuple):
-> 1141             card = Card(*card)
   1142         elif card is None:
   1143             card = Card()

~/lsst/lsstsw/miniconda/envs/lsst-scipipe/lib/python3.7/site-packages/astropy/io/fits/card.py in __init__(self, keyword, value, comment, **kwargs)
    189                 self.keyword = keyword
    190             if value is not None:
--> 191                 self.value = value
    192 
    193         if comment is not None:

~/lsst/lsstsw/miniconda/envs/lsst-scipipe/lib/python3.7/site-packages/astropy/io/fits/card.py in value(self, value)
    309                            np.floating, np.integer, np.complexfloating,
    310                            np.bool_)):
--> 311             raise ValueError('Illegal value: {!r}.'.format(value))
    312 
    313         if isinstance(value, float) and (np.isnan(value) or np.isinf(value)):

ValueError: Illegal value: <Quantity 10. deg_C>.
```

I would have expected to be able to insert a Quantity into the header and have the units round-trip. The FITS standard document suggests putting units in the comment field for that header key (see section 4.3.2 of the FITS Standard document); since astropy Quantities have well defined stringifications, we could just use those?

**Comment by pllim** [comment]
Duplicate of #9197?

**Comment by parejkoj** [comment]
@pllim : I don't think so, no. #9197 is about writing to a FITS BinTableHDU (which is already [partly supported via mixins](https://docs.astropy.org/en/stable/io/unified.html#astropy-native-objects-mixin-columns)). This one is about writing to the FITS headers.

**Comment by pllim** [comment]
So, what did you expect to appear in the header if things Just Works ™️ ?

**Comment by timj** [comment]
Something like:
```
CCDTEMP  =     10 / [C]
```
(modulo me getting the spacing right and remembering how FITS was Celsius to be written). The unit should be inserted at the start of any existing comment (overwriting unit if one is already defined). Parsing the units out on when extracting values from a header would also be wonderful.

**Comment by MSeifert04** [comment]
I'm a bit worried that this will be a bit too much magic for most users and it could well be a maintenance nightmare. For example with CCDData unit support when reading from filte files was a frequent issue - and it only used the BUNIT keyword value).

**Comment by mhvk** [comment]
I'd love to have this! And I think it is enough of a practice that one could just look for anything within `[...]` and try to parse as a unit.
Alternatively, more astropy specific, we could perhaps ensure it gets into the a special commented section, like we do for mixin-columns.
Ideally, though, we follow some form of standard here...

**Comment by jpmorgen** [comment]
Executive summary: I have solved this problem in a configurable way and am wondering if the community is still interested.  I am new to the community, so I need some help deciding whether my solution is best as (1) something on my personal github site (2) a non-astropy PyPy package (3) an astropy affiliate package  (4) part of astropy.  Thanks in advance for any guidance.

Details:

Regarding this thread, I am creating the mixin FitsKeyQuantity.  It can using regex magic and some property, it can put/look for the unit specification at the beginning and end of the FITS card comment enclosed however you want (location and delimeters could of course be standardized).  With the help of this thread, the issue I identified as potentially confusing for end-users is whether or not general queries to FITS keys (e.g. value = ccd.meta['KEYWORD']) will return Quantities or values.  My idea to address this is to have a property fits_key_return_as_quantity with possible values (1) 'never': return numeric values only; (2) 'recognized': return Quantites when identified as such and (3) 'always': return Quantities with units of dimensionless_unscaled when units were otherwise not found.  The default would be 'never.'  These choices could also be passed to meta.get() as a keyword so as to override the property when you need to.

I have also created the mixin FitsKeyArithmetic which enables numeric fits keys to participate in NDArithmetic.  I use this to track the CCD saturation level and nonlinearity point through the reduction and analysis process.  The system is opt-in -- simply provide a list of keys that you want to have participate and it does the magic.  The only sure calculation is when NDArithmetic operand2 is a scalar, so in general I just give up and delete the key when operand2 is an image.  However, it is possible to pass in a function, `handle_image.` (thanks for that example in NDArithmetic!)  For instance, for biases and darks, my handle_image function returns 0 as the scalar distillation of operand2. 

Regarding BUNIT --> ccd.unit, as a new user (coming from years with IDL), I did struggle with that -- I was just trying to get my *** data into the CCDData object!  I have suggested some solutions that would have made that transition easier for this test particle (PR #11325).  BTW learning how to subclass an object with a "read" class method was REALLY challenging.  But challenge breeds progress!

**Comment by mhvk** [comment]
@jpmorgen - that sounds great!  I think having a configurable way to deal (or not deal) with units on header cards sounds like the way to go. It would be really nice to see some explicit examples - as a pull request is OK (would help to try to split it up to the extent possible).

Note that our 4.3 feature freeze is almost here, so the next week responses may be a little slow...

**Comment by jpmorgen** [comment]
The attached file implements two draft PRs as subclasses of existing astropy objects.  The file includes extensive test/demonstration code in __main__.  The file is also available at https://github.com/jpmorgen/astropy_fits_key.  

===============
PR1: QuantityCard, QuantityHeader

(1) Adds `unit` property to Cards

(2) `unit` property can be set directly, using Card.value as a Quantity, or by inserting an appropriately formatted unit string to Card.comment

(2a) Setting Card.value = Quantity ALWAYS does the right thing, since it sets Card.unit and adds a unit string to Card.comment

(3) Default behavior of extraction of value as Quantity is controlled using `return_key_as_quantity` property

(3a) Accessing Card.unit always works (or returns None) and therefore may end up being preferred way of reliably constructing the Quantity of a card.  The Card.unit property could be used at the Header level to implement something like a "Header.get_quantity(key)" method

(4) The presentation of unit.to_string() within the Card.comment string is controlled by property which currently defaults to:

unit_str_start = ' '
unit_str_delimiters = '()'
unit_str_end = ''
unit_str_position = 'end'

This puts a unit string such as " (s)" on the end of a comment.  I think this looks better when there is other explanatory text in the comment.  However, when there is no text at all, putting it at the beginning as recommended by @timj looks better.  I am totally open to the community vote on that.

PR1 TODO: improve integration with Header (e.g. 3a).

=======================

PR2: FitsKeyArithmeticMixin

This PR enables NDArithmetic-style arithmetic on fits keys listed in the `arithmetic_keylist` property.  Handles Card units if present, otherwise defaults to assuming cards have the same unit as the invoking CCDData.

TODO: Resolve angst about what to do when user (or underlying routines user is using, like ccdproc.gain_correct) doesn't specify handle_meta="first_found" in call to ccd.<arithmetic_operation> or otherwise make the CCDData meta available in NDData._arithmetic.  Currently, specification of `arithmetic_keylist` always triggers handle_meta="first_found"-like behavior.


[fits_key.py.txt](https://github.com/astropy/astropy/files/6418659/fits_key.py.txt)

**Comment by saimn** [comment]
@jpmorgen - your first patch (PR1) seems close to what we would like here, but it's a bit difficult to parse the changes in the current format. Would you mind making this a real PR where I and other interested people could review the code and make suggestions ?

**Comment by jpmorgen** [comment]
@saimn thanks for the encouragement 

PR #11849 contains draft code for PR1

NEW: Header.quantity method.  This always returns the card value as a Quantity or raises an error if a Quantity cannot be constructed.  This should hopefully resolve angst about whether or not to return card values as Quantity when accessed through the standard methods (e.g. hdr['KEYWORD']): the standard methods work the way they always did

OLD: I still have angst about whether to standardize the placement and format of the unit strings or not.  Parenthetical comments at the ends of comments raise warnings, which are turned into errors in pytest

Test code is attached and available at [https://github.com/jpmorgen/astropy_fits_key]()

[card_quantity_tester.txt](https://github.com/astropy/astropy/files/6657202/card_quantity_tester.txt)
